#Coding the Likelihood Function



*   Conditional Likelihood
*   Unconditional Likelihood



In [ ]:
import numpy as np
from scipy.stats import norm

def ar7_conditional_likelihood(params, data):
    """
    Calculate the conditional likelihood of an AR(7) model.

    Parameters:
        params (list): Parameters [phi1, phi2, ..., phi7, sigma^2].
        data (array): Array of observed data.

    Returns:
        float: Log likelihood value.
    """
    phi = params[:-1]
    sigma_squared = params[-1]
    n = len(data)
    ll = 0

    # First 7 observations are given, likelihood starts from the 8th observation
    for t in range(7, n):
        # Calculate mean based on past 7 observations
        mean = np.dot(phi, data[t-7:t][::-1])
        # Calculate log likelihood contribution of current observation
        ll += norm.logpdf(data[t], loc=mean, scale=np.sqrt(sigma_squared))

    return ll

def ar7_unconditional_likelihood(params, data):
    """
    Calculate the unconditional likelihood of an AR(7) model.

    Parameters:
        params (list): Parameters [phi1, phi2, ..., phi7, sigma^2].
        data (array): Array of observed data.

    Returns:
        float: Log likelihood value.
    """
    phi = params[:-1]
    sigma_squared = params[-1]
    n = len(data)
    ll = 0

    # Calculate the unconditional mean of the AR(7) process
    mu = np.dot(phi, data[:7][::-1])

    # Calculate log likelihood for each observation using unconditional mean
    for t in range(n):
        ll += norm.logpdf(data[t], loc=mu, scale=np.sqrt(sigma_squared))

    return ll

# Example usage:
# Define some sample data and initial parameters
data = np.random.normal(size=100)  # Sample data
initial_params = [0.5, -0.2, 0.3, -0.1, 0.2, -0.4, 0.1, 1.0]  # Initial parameters

# Calculate conditional likelihood
conditional_ll = ar7_conditional_likelihood(initial_params, data)
print("Conditional Log Likelihood:", conditional_ll)

# Calculate unconditional likelihood
unconditional_ll = ar7_unconditional_likelihood(initial_params, data)
print("Unconditional Log Likelihood:", unconditional_ll)


Conditional Log Likelihood: -155.24845520747374
Unconditional Log Likelihood: -132.22901613926624


**Conditional Log Likelihood: -155.24845520747374**

The conditional log likelihood (-155.24845520747374) is calculated based on the assumption that the model's predictions are conditioned on the previous 7 observations.
This means that for each data point beyond the 7th observation, the model's prediction is based on the previous 7 observations.
A higher conditional log likelihood indicates that the model's predictions align well with the observed data given this conditioning.
The conditional log likelihood value suggests that the AR(7) model, with the provided initial parameters, provides a reasonable fit to the observed data when conditioning on the previous 7 observations.

**Unconditional Log Likelihood: -132.22901613926624**
The unconditional log likelihood (-132.22901613926624) is calculated assuming that the model's predictions are based on a constant unconditional mean computed from the first 7 observations. This means that the model's predictions are not conditioned on any specific past observations beyond the first 7. A higher unconditional log likelihood indicates that the model's predictions provide a good fit to the entire dataset, without conditioning on specific past observations.The unconditional log likelihood value suggests that the AR(7) model, with the provided initial parameters, also provides a reasonable fit to the observed data when assuming a constant unconditional mean.

#Maximizing the Likelihood



In [ ]:
from scipy.optimize import minimize

# Define negative log likelihood functions (to be minimized)
def neg_log_likelihood_conditional(params, data):
    return -ar7_conditional_likelihood(params, data)

def neg_log_likelihood_unconditional(params, data):
    return -ar7_unconditional_likelihood(params, data)

# Example usage:
# Define some sample data
data = np.random.normal(size=100)  # Sample data

# Initial parameter guess
initial_params = [0.5, -0.2, 0.3, -0.1, 0.2, -0.4, 0.1, 1.0]

# Maximize conditional likelihood
result_conditional = minimize(neg_log_likelihood_conditional, initial_params, args=(data,), method='Nelder-Mead')
print("Maximized Conditional Likelihood Parameters:", result_conditional.x)

# Maximize unconditional likelihood
result_unconditional = minimize(neg_log_likelihood_unconditional, initial_params, args=(data,), method='Nelder-Mead')
print("Maximized Unconditional Likelihood Parameters:", result_unconditional.x)


Maximized Conditional Likelihood Parameters: [ 0.07306005 -0.0252478   0.02007128 -0.12153876  0.07980201 -0.05816385
 -0.09271792  0.94545767]
Maximized Unconditional Likelihood Parameters: [-0.10258685 -0.2644132   0.42631916 -0.13179205  0.20967231 -0.40713626
  0.09187048  0.95030504]


**Maximized Conditional Likelihood Parameters:**


*   The parameters [ 0.07306005, -0.0252478, 0.02007128, -0.12153876, 0.07980201, -0.05816385, -0.09271792, 0.94545767] are the estimates that maximize the conditional likelihood function.


*   These parameters are obtained by minimizing the negative log likelihood of the conditional likelihood function using the Nelder-Mead optimization method.
*   They represent the coefficients (phi) and the variance (sigma^2) of the AR(7) model that best fit the observed data when conditioned on the previous 7 observations.


Maximized Unconditional Likelihood Parameters: **bold text**


*   The parameters [-0.10258685, -0.2644132, 0.42631916, -0.13179205, 0.20967231, -0.40713626, 0.09187048, 0.95030504] are the estimates that maximize the unconditional likelihood function.
*   These parameters are obtained by minimizing the negative log likelihood of the unconditional likelihood function using the Nelder-Mead optimization method.

*   They represent the coefficients (phi) and the variance (sigma^2) of the AR(7) model that best fit the observed data without conditioning on specific past observations.

#Parameter Estimation

In [ ]:
import numpy as np
import pandas as pd
from scipy.optimize import minimize

# Step 1: Data Preprocessing
# Load FRED-MD dataset
data = pd.read_csv('/content/current.csv')  # Replace 'path_to_your_dataset.csv' with the actual path

# Assuming 'INDPRO' is the column name for the INDPRO variable
indpro = data['INDPRO']

# Calculate monthly log differences
log_diff_indpro = np.log(indpro).diff().dropna()

# Step 2: Likelihood Functions
def ar7_conditional_likelihood(params, data):
    phi, sigma_sq = params[:7], params[7]
    n = len(data)
    ll = 0
    for t in range(7, n):
        yt = data[t]
        mu_t = np.dot(phi, data[t-7:t][::-1])  # Reverse the order to align with AR(7) formula
        ll += -0.5 * (np.log(2 * np.pi * sigma_sq) + (yt - mu_t)**2 / sigma_sq)
    return ll

def ar7_unconditional_likelihood(params, data):
    phi, sigma_sq = params[:7], params[7]
    n = len(data)
    mu0 = np.mean(data[:7])  # Initial mean
    ll = -0.5 * (7 * np.log(2 * np.pi * sigma_sq) + np.sum((data[:7] - mu0)**2) / sigma_sq)
    for t in range(7, n):
        yt = data[t]
        mu_t = np.dot(phi, data[t-7:t][::-1])  # Reverse the order to align with AR(7) formula
        ll += -0.5 * (np.log(2 * np.pi * sigma_sq) + (yt - mu_t)**2 / sigma_sq)
    return ll

# Step 3: Parameter Estimation
# Initial parameter guess
initial_params = np.array([0.5, -0.2, 0.3, -0.1, 0.2, -0.4, 0.1, 1.0])

# Maximize conditional likelihood
result_conditional = minimize(ar7_conditional_likelihood, initial_params, args=(log_diff_indpro,), method='Nelder-Mead')
print("Maximized Conditional Likelihood Parameters:", result_conditional.x)

# Maximize unconditional likelihood
result_unconditional = minimize(ar7_unconditional_likelihood, initial_params, args=(log_diff_indpro,), method='Nelder-Mead')
print("Maximized Unconditional Likelihood Parameters:", result_unconditional.x)


Maximized Conditional Likelihood Parameters: [-1.05849043e+50  1.06746350e+49 -1.73462824e+49  2.64622188e+49
  2.68764340e+49 -4.69880006e+49 -3.21491762e+49  6.97645179e+50]
Maximized Unconditional Likelihood Parameters: [-1.05849043e+50  1.06746350e+49 -1.73462824e+49  2.64622188e+49
  2.68764340e+49 -4.69880006e+49 -3.21491762e+49  6.97645179e+50]


The extremely large values obtained for the parameter estimates in both the conditional and unconditional likelihood maximizations suggest that the optimization process did not converge to meaningful solutions.

#Forecasting

In [ ]:
import numpy as np

# Function to forecast future values
def forecast_ar7(params, data, steps):
    # Extract parameters
    phi = params[:7]
    sigma = params[7]

    # Initialize forecast array with observed data
    forecast = np.copy(data)

    # Forecast future values
    for i in range(steps):
        forecast_next = np.dot(phi, forecast[-7:]) + np.random.normal(0, sigma)
        forecast = np.append(forecast, forecast_next)

    return forecast

# Forecast using conditional likelihood parameters
conditional_forecast = forecast_ar7(result_conditional.x, log_diff_indpro, 8)
print("Conditional Forecast:", conditional_forecast[-8:])

# Forecast using unconditional likelihood parameters
unconditional_forecast = forecast_ar7(result_unconditional.x, log_diff_indpro, 8)
print("Unconditional Forecast:", unconditional_forecast[-8:])


Conditional Forecast: [-4.85355126e+050  1.56037675e+100 -5.01648269e+149  1.61275786e+199
 -5.18488365e+248  1.66689738e+298             -inf              nan]
Unconditional Forecast: [-7.80827157e+050  2.51029498e+100 -8.07039157e+149  2.59456440e+199
 -8.34131081e+248  2.68166271e+298             -inf              nan]


The forecasted values obtained from both the conditional and unconditional AR(7) models exhibit extremely large and divergent values, as well as non-finite values such as inf and nan. This indicates that the forecasting process did not produce meaningful results.